CRFModuleValidation/MasterScheduledActivityManager.swift

```diff
@@ -0,0 +1,563 @@
+    
+    override init(delegate: SBAScheduledActivityManagerDelegate?) {
+        super.init(delegate: delegate)
+        
+        // Set days behind and days ahead to only cache today's activities
+        self.daysBehind = 0
+        self.daysAhead = 15
+    }
+    
+    func resetScheduleFilter() {
+        let todayStart = Date().startOfDay() as NSDate
+        let finishedTodayOrFuture = NSPredicate(format: "finishedOn == nil OR finishedOn >= %@", todayStart)
+        let expiresTodayOrFuture = NSPredicate(format: "expiresOn == nil OR expiresOn >= %@", todayStart)
+        self.scheduleFilterPredicate = NSCompoundPredicate(andPredicateWithSubpredicates: [finishedTodayOrFuture, expiresTodayOrFuture])
+    }
+    
+    func completedCount(for taskGroup: TaskGroup) -> Int {
+        return scheduleSections.filter({ $0.items.find({ $0.taskGroup == taskGroup })?.isCompleted ?? false }).count
+    }
+    
+    func isTodayCompleted(for taskGroup: TaskGroup) -> Bool {
+        guard let todaySection = self.scheduleSections.first(where: { Calendar.gregorian.isDateInToday($0.date) }),
+            let todayItem = todaySection.items.first(where: { $0.taskGroup == taskGroup })
+        else {
+            return false
+        }
+        return todayItem.isCompleted
+    }
+    
+    func schedule(for taskGroup: TaskGroup) -> Schedule? {
+        return self.schedules.first(where: { $0.taskGroup == taskGroup })
+    }
+    
+    func shouldFireTimingSchedule(for taskGroup:TaskGroup) -> Bool {
+        guard let schedule = schedule(for: taskGroup)
+        else {
+            return false
+        }
+        return !schedule.hasData && completedCount(for: taskGroup) <= 1
+    }
+    
+    override func resetData() {
+        scheduleSections.removeAll()
+        schedules.removeAll()
+        dayOne = nil
+    }
+    
+    override func reloadData() {
+        // Unless today's date has changed, rely upon the previously loaded data
+        guard !Calendar.current.isDateInToday(today) || scheduleSections.count == 0 else { return }
+        
+        forceReload()
+    }
+    
+    func forceReload() {
+    
+        // Exit early if loading
+        guard !self.isReloading else { return }
+        
+        // Mark the date as today and reset the filter
+        resetScheduleFilter()
+        today = Date()
+        
+        // load schedules
+        loadScheduledActivities(from: startStudy, to: scheduleAhead)
+    }
+    
+    override func load(scheduledActivities: [SBBScheduledActivity]) {
+
+        // Save the clinic scheduled activity for this user's data groups
+        guard let dataGroups = SBAUser.shared.dataGroups else { return }
+        let clinics = dataGroups.mapAndFilter({ (group) -> String? in
+            if group.hasPrefix("clinic") { return group }
+            return nil
+        })
+        guard clinics.count == 1 else { return }
+        let clinicIdentifier = clinics.first!
```

Comment:
````
guard let dataGroups = SBAUser.shared.dataGroups,
        let clinicIdentifier = dataGroups.first(where: { $0.hasPrefix("clinic") }) 
    else { 
        return 
}
````

Data groups is already a string so you don't need to map and filter it. Also, you are only looking for one. either clinic1 or clinic2 so you can just return the first one found.

Reply:
Thanks, I'm going with the excuse that it was getting late and I was feeling rushed.